### ISSUE TITLE
Refer to a value created in dynamics

### ISSUE THREAD

**Issue by axozoid** [sparkleformation/sparkle_formation#263]
HI there,
I've stuck in trying to figure out how to access a return value of a newly created resource?
For example, there is a code:
``` 
vpc_endpoint = dynamic!(:aws_ec2_vpc_endpoint, :test_vpc) do
      properties do
        service_name svc[:service_name]
        vpc_endpoint_type 'Interface'
        vpc_id vpc_id
        subnet_ids svc[:subnet_ids]
        security_group_ids [ref!(security_group.resource_name!)]
      end # properties do
    end # dynamic do
```

My next step is to create a nice Route53 record which will be an alias to the VPC endpoint created with the above code. I'm trying to do something like this, but that's not working:
```
 dynamic!(
       :record,
       :test_alias,
       name: 'test-alias-record.example.com',
       hosted_zone_name: 'example.com',
       type: 'A',
       alias_target: { # should be the reference to the endpoint created above
         d_n_s_name: ref!(vpc_endpoint.DnsEntries[0].DnsName!),
         hosted_zone_id: ref!(vpc_endpoint.DnsEntries[0].HostedZoneId!)
       }
     )
```
I assume there should be something like [`GetAtt`](https://docs.aws.amazon.com/en_pv/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html) but can't find any examples of how to use it.

Reference links:
- [AWS VPC endpoint properties and return values](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-vpcendpoint.html);

Please assist!
Many thanks!

**Comment by axozoid** [comment]
If I do
```
puts ref!(vpc_endpoint.resource_name!)
```
It shows me the proper name of the resource, but can't get below and access the return values.

**Comment by chrisroberts** [comment]
Hi! SparkleFormation provides helper methods for the intrinsic functions within CloudFormation. In this case, you would access the resource attributes via the `Fn::GetAttr` intrinsic function, and SparkleFormation provides a helper for this as `attr!`.

In the example you provide above, you'll need to use a couple different intrinsic functions to get to the actual data you're after. For the list of dns entries, you would use:

```ruby
attr!(:vpc_endpoint, :dns_entries)
```

This returns the array of entries, but if you only want the first, you'll need to use `select!` to extract it:

```ruby
select!(0, attr!(:vpc_endpoint, :dns_entries))
```

The single entry consists of two parts, the hosted zone ID, and the DNS. You'll need to extract this information for each part using `split!` and `select!` so it should end up looking something like:

```ruby
d_n_s_name select!(1, split!(":", select!(0, attr!(:vpc_endpoint, :dns_entries))))
hosted_zone_id select!(0, split!(":", select!(0, attr!(:vpc_endpoint, :dns_entries))))
```

Hope that helps. Cheers!

**Comment by axozoid** [comment]
hi @chrisroberts ! thank you very much for a such detailed explanation! I'm sure this will help others to to get more understanding about Sparkleformation.

Coming back to the code in this thread.

When I do `sfn print` I get this:
```
{
  "Description": "SparkleFormation Stack",
  "Resources": {
    "PrivateLinkSecurityGroupAwsEc2SecurityGroup": {
      "Type": "AWS::EC2::SecurityGroup",
      "Properties": {
        "GroupDescription": "Private Link Security Group",
        "VpcId": "vpc-09c60c6d",
        "SecurityGroupIngress": {
          "Description": "Private Link - Allow all traffic",
          "IpProtocol": "tcp",
          "CidrIp": "0.0.0.0/0",
          "FromPort": 0,
          "ToPort": 65535
        }
      }
    },
    "PreviewcontentapiAwsEc2VpcEndpoint": {
      "Type": "AWS::EC2::VPCEndpoint",
      "Properties": {
        "ServiceName": "<REDACTED>",
        "VpcEndpointType": "Interface",
        "VpcId": "vpc-09c60c6d",
        "SubnetIds": [
          "subnet-f7be0781",
          "subnet-71f56415"
        ],
        "SecurityGroupIds": [
          {
            "Ref": "PrivateLinkSecurityGroupAwsEc2SecurityGroup"
          }
        ]
      }
    },
    "TestAliasRecordSet": {
      "Type": "AWS::Route53::RecordSet",
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::Select": [
              1,
              {
                "Fn::Split": [
                  {
                    "Fn::Select": [
                      0,
                      {
                        "Fn::GetAtt": [
                          "VpcEndpoint",
                          "DnsEntries"
                        ]
                      }
                    ]
                  },
                  ":"
                ]
              }
            ]
          },
          "HostedZoneId": {
            "Fn::Select": [
              0,
              {
                "Fn::Split": [
                  {
                    "Fn::Select": [
                      0,
                      {
                        "Fn::GetAtt": [
                          "VpcEndpoint",
                          "DnsEntries"
                        ]
                      }
                    ]
                  },
                  ":"
                ]
              }
            ]
          }
        },
        "Comment": {
          "Ref": "AWS::NoValue"
        },
        "Failover": {
          "Ref": "AWS::NoValue"
        },
        "HealthCheckId": {
          "Ref": "AWS::NoValue"
        },
        "HostedZoneId": {
          "Ref": "AWS::NoValue"
        },
        "HostedZoneName": "<REDACTED>",
        "Name": "<REDACTED>",
        "Region": {
          "Ref": "AWS::NoValue"
        },
        "ResourceRecords": {
          "Ref": "AWS::NoValue"
        },
        "SetIdentifier": {
          "Ref": "AWS::NoValue"
        },
        "TTL": {
          "Ref": "AWS::NoValue"
        },
        "Type": "A",
        "Weight": {
          "Ref": "AWS::NoValue"
        }
      }
    }
  },
  "Outputs": {
    "TestAliasName": {
      "Description": "Domain name of record set.",
      "Value": {
        "Ref": "TestAliasRecordSet"
      }
    }
  }
}

```
Which seems OK so far, but when I try to apply it by `sfn create`, I get:
```
[Sfn]: SparkleFormation: create
[Sfn]:   -> Name: privatelink
ERROR: Miasma::Error::ApiError::RequestError: Bad Request - ValidationError: Template error: instance of Fn::GetAtt references undefined resource VpcEndpoint
```

This is my dynamic to create a DNS record:
```
    dynamic!(
      :record,
      :test_alias,
      name: "#{svc_name}.#{_hosted_zone_name}",
      hosted_zone_name: "#{_hosted_zone_name}",
      type: 'A',
      alias_target: {
        d_n_s_name: select!(1, split!(":", select!(0, attr!(:vpc_endpoint, :dns_entries)))),
        hosted_zone_id: select!(0, split!(":", select!(0, attr!(:vpc_endpoint, :dns_entries))))
      }
    )
```

Do you have any ideas why it's complaining?
Thank you in advance!

**Comment by axozoid** [comment]
Update. I managed to get it working.
When I used `attr!(:vpc_endpoint, :dns_entries)` it didn't perform a resource lookup and I had this in my CF code:
```
                        "Fn::GetAtt": [
                          "VpcEndpoint",
                          "DnsEntries"
                        ]
```
When I replaced `attr!(:vpc_endpoint, :dns_entries)` with `attr!(vpc_endpoint.resource_name!, 'DnsEntries')` I started seeing reference to a proper resource and not to `VpcEndpoint`.

Next issue discovered, is that `split` wasn't working as expected. 
For example, using `split!(":", "123:456")` was resulting as `"Fn::Split": ["123:456", ":"]` in CF code, which is not the right format ([ref link](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-split.html)). 
Flipping the arguments helped to get the right order and resulting code became matching what was expecting by the CF. Since then `sfn create` worked as expected.
